### PR TITLE
Github actions for Az cli, login and ARM deployment

### PR DESCRIPTION
**Pull request by dikhakha**

### CODE REVIEW

File: azure-template-deployment/entrypoint.sh [hunk 1]
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash
+
+set -e
+
+az group create --name "$RESOURCE_GROUP_NAME" --location "$RESOURCE_GROUP_LOCATION"
```
Comment (ds-ms) [inline azure-template-deployment/entrypoint.sh]: Will this not fail, if the RG already exists?

File: azure-template-deployment/entrypoint.sh [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+URI_REGEX="^(http://|https://)\\w+"
+
+if [[ $TEMPLATE_FILE =~ $URI_REGEX ]]
+then
+  az group deployment create -g "$RESOURCE_GROUP_NAME" --name "$DEPLOYMENT_NAME" --template-uri "$TEMPLATE_FILE" --parameters "@${GITHUB_WORKSPACE}/${PARAMETERS_FILE_PATH}"
```
Comment (ds-ms) [inline azure-template-deployment/entrypoint.sh]: I suggest you generate the `$DEPLOYMENT_NAME`, otherwise the user would have to update the workflow everytime (s)he wants to deploy

File: azure-template-deployment/entrypoint.sh (hunk 1 above)
Comment (dikhakha) [reply azure-template-deployment/entrypoint.sh]: no it is a no op if RG exists

File: azure-template-deployment/entrypoint.sh (hunk 2 above)
Comment (dikhakha) [reply azure-template-deployment/entrypoint.sh]: got it. i can have a default

Comment (dikhakha) [reply azure-template-deployment/entrypoint.sh]: but why would an update require for deployment name ? i think i need not be unique. will check

File: azure-login/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+
+LABEL version="1.0.0"
+
+LABEL maintainer="Disha Khakhar"
+LABEL com.github.actions.name="Azure Login"
+LABEL com.github.actions.description="GitHub Action to login in to Azure"
+LABEL com.github.actions.icon="triange"
+LABEL com.github.actions.color="blue"
+
+ENTRYPOINT az login --service-principal --username "$AZURE_SERVICE_APP_ID" --password "$AZURE_SERVICE_PASSWORD" --tenant "$AZURE_SERVICE_TENANT"
```
Comment (vincentdass) [inline azure-login/Dockerfile]: suggestion: Can u keep a separate entrypoint.sh file for this and check if the required variables are non empty.

File: azure-cli/entrypoint.sh [hunk 4]
```diff
@@ -0,0 +1,11 @@
+#!/bin/sh
+
+set -e
+
+if [ -n "$SCRIPT_PATH" ]
+then
+  chmod +x ${GITHUB_WORKSPACE}/${SCRIPT_PATH}
```
Comment (sachinma) [inline azure-cli/entrypoint.sh]: Should we check if the file exists and throw a better message if that is not the case

File: azure-login/Dockerfile (hunk 3 above)
Comment (sachinma) [reply azure-login/Dockerfile]: +1. You can have better messages if the required parameters are missing.

File: azure-template-deployment/entrypoint.sh (hunk 1 above)
Comment (dikhakha) [reply azure-template-deployment/entrypoint.sh]: i have added check for RG creation

File: azure-template-deployment/entrypoint.sh (hunk 2 above)
Comment (dikhakha) [reply azure-template-deployment/entrypoint.sh]: i have added GUID to the deployment name if not present

File: azure-login/Dockerfile (hunk 3 above)
Comment (dikhakha) [reply azure-login/Dockerfile]: done

File: azure-cli/entrypoint.sh (hunk 4 above)
Comment (dikhakha) [reply azure-cli/entrypoint.sh]: done

File: arm/entrypoint.sh [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+URI_REGEX="^(http://|https://)\\w+"
+GUID=$(uuidgen | cut -d '-' -f 1)
+
+# Download parameters file if it is a remote URL
+
+if [[ $AZURE_TEMPLATE_PARAM_FILE =~ $URI_REGEX ]]
+then
+  PARAMETERS_FILE="parameters-${GUID}.json"
+  TEMPLATE_PARAMETERS=$(curl $AZURE_TEMPLATE_PARAM_FILE)
+  echo ${TEMPLATE_PARAMETERS} >> $PARAMETERS_FILE
```
Comment (ds-ms) [inline arm/entrypoint.sh]: You could also do `curl $AZURE_TEMPLATE_PARAM_FILE -o $PARAMETERS_FILE`

File: arm/entrypoint.sh [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+  echo "Generated Deployment Name ${DEPLOYMENT_NAME}"
+fi
+
+# Deploy ARM template
+
+if [[ $AZURE_TEMPLATE_LOCATION =~ $URI_REGEX ]]
+then
+  az group deployment create -g "$AZURE_RESOURCE_GROUP" --name "$DEPLOYMENT_NAME" --template-uri "$AZURE_TEMPLATE_LOCATION" --parameters "@$PARAMETERS_FILE"
+else
+  TEMPLATE_FILE="${GITHUB_WORKSPACE}/${AZURE_TEMPLATE_LOCATION}"
```
Comment (ds-ms) [inline arm/entrypoint.sh]: Does this not work if you don't suffix `$GITHUB_WORKSPACE`; 
I suppose the action is being executed in the workspace, so the relative path should work?

File: arm/entrypoint.sh (hunk 5 above)
Comment (dikhakha) [reply arm/entrypoint.sh]: super ! that works. ill update it.

File: arm/entrypoint.sh (hunk 6 above)
Comment (dikhakha) [reply arm/entrypoint.sh]: ill check if thats the case

File: arm/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+
+LABEL version="1.0.0"
+
+LABEL maintainer="Disha Khakhar"
+LABEL com.github.actions.name="Manage Azure Resources"
+LABEL com.github.actions.description="GitHub Action to deploy ARM template to Azure"
+LABEL com.github.actions.icon="triange"
+LABEL com.github.actions.color="blue"
+
+RUN apk update \
```
Comment (sachinma) [inline arm/Dockerfile]: Why is this required?

File: arm/entrypoint.sh [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+set -e
+
+# Create Resource group if it does not exists
+
+RESOURCE_GROUP_EXISTS=$(az group exists -n "$AZURE_RESOURCE_GROUP")
+if [[ $RESOURCE_GROUP_EXISTS == "false" ]]
+then
+  if [[ -z "$RESOURCE_GROUP_LOCATION" ]]
+  then
+    echo "RESOURCE_GROUP_LOCATION is not set."
```
Comment (sachinma) [inline arm/entrypoint.sh]: You should exit after this line

File: arm/entrypoint.sh [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+  else
+    az group create --name "$AZURE_RESOURCE_GROUP" --location "$RESOURCE_GROUP_LOCATION"
+  fi
+fi
+
+URI_REGEX="^(http://|https://)\\w+"
```
Comment (sachinma) [inline arm/entrypoint.sh]: \w will it work with more slashes or . etc?

File: arm/entrypoint.sh (hunk 5 above)
Comment (sachinma) [inline arm/entrypoint.sh]: Do you need to create the file? Why not just have it in memory?

File: azure-login/entrypoint.sh [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+if [[ -n "$AZURE_SERVICE_APP_ID" ]] && [[ -n "$AZURE_SERVICE_PASSWORD" ]] && [[ -n "$AZURE_SERVICE_TENANT" ]]
+then
+  az login --service-principal --username "$AZURE_SERVICE_APP_ID" --password "$AZURE_SERVICE_PASSWORD" --tenant "$AZURE_SERVICE_TENANT"
+else
+  echo "One of the required parameters for Azure Login is not set: AZURE_SERVICE_APP_ID, AZURE_SERVICE_PASSWORD, AZURE_SERVICE_TENANT."
+  exit 1
+fi
+
+SUBSCRIPTIONS=$(az account list)
```
Comment (sachinma) [inline azure-login/entrypoint.sh]: Should you get  the list ony if $Azure_subscription is not provided

Comment (dikhakha) [reply azure-login/entrypoint.sh]: got it.

File: arm/entrypoint.sh (hunk 5 above)
Comment (dikhakha) [reply arm/entrypoint.sh]: Because if it is a file, the command format is --parameters @param_file_name , for inline it is --parameters 'params'. However i can handle this if condition. will do.

File: arm/Dockerfile (hunk 7 above)
Comment (dikhakha) [reply arm/Dockerfile]: it is used to get the list of latest available packages, so this line is usually added just before apk add package

File: arm/entrypoint.sh (hunk 9 above)
Comment (dikhakha) [reply arm/entrypoint.sh]: it wont match the whole url as / and . will be ignored. we can come up with a more robust url regex. the check will pass for anything with prefix http://|https:// followed by a character followed by any set of special | non special characters

File: arm/entrypoint.sh (hunk 8 above)
Comment (dikhakha) [reply arm/entrypoint.sh]: good catch, will do

File: arm/Dockerfile (hunk 7 above)
Comment (sachinma) [reply arm/Dockerfile]: I meant both the lines. Ok do we need util-linux? What for

File: azure-login/entrypoint.sh [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+fi
+
+if [[ -n "$AZURE_SUBSCRIPTION" ]]
+then
+  az account set --s "$AZURE_SUBSCRIPTION"
+else
+  SUBSCRIPTIONS=$(az account list)
+  if [[ ${#SUBSCRIPTIONS[@]} > 1 ]]
+  then
+    echo "AZURE_SUBSCRIPTION is not set."
```
Comment (sachinma) [inline azure-login/entrypoint.sh]: Do you need to write your error statements to error stream throughout (> 2 ) to mark them as errors

File: arm/Dockerfile (hunk 7 above)
Comment (dikhakha) [reply arm/Dockerfile]: for generation of guid for deployment name. uuidgen is part of util-linux package

File: azure-login/entrypoint.sh (hunk 11 above)
Comment (dikhakha) [reply azure-login/entrypoint.sh]: will do.

**Review by sachinma — APPROVED** [review]
